### PR TITLE
Implement smoothing for fonts

### PR DESCRIPTION
**Pull request by JonnyPtn**
## Description

Following on from forum discussion [here](https://en.sfml-dev.org/forums/index.php?topic=21058.0) this implements smoothing for fonts, via `setSmooth();`

## Tasks

* [x] Tested on Linux
* [x] Tested on Windows
* [x] Tested on macOS
* [ ] Tested on iOS
* [ ] Tested on Android

## How to test this PR?

Can test using the following code. Space toggles the font's smoothing

```cpp

#include <SFML/Graphics.hpp>
int main()
{

    sf::RenderWindow window( sf::VideoMode( 800, 600, 32 ), "SFML text smoothing",
                             sf::Style::Titlebar | sf::Style::Close );


    sf::Font font;
    if (!font.loadFromFile("resources/font.ttf"))
        return EXIT_FAILURE;

    sf::Text message;
    message.setFont(font);
    int textSize = 40;
    message.setScale( { 1.5f,1.5f } );
    message.setCharacterSize(textSize);
    message.setFillColor(sf::Color::White);

    message.setString( "Some text about something or other" );
    

    while (window.isOpen())
    {
        // Handle events
        sf::Event event;
        while ( window.pollEvent( event ) )
        {
            // Window closed or escape key pressed: exit
            if ( (event.type == sf::Event::Closed) ||
                 ((event.type == sf::Event::KeyPressed) && (event.key.code == sf::Keyboard::Escape)) )
            {
                window.close();
                break;
            }

            if ( event.type == sf::Event::KeyPressed )
            {
                switch ( event.key.code )
                {
                case sf::Keyboard::Space:
                    font.setSmooth( !font.isSmooth() );
                    break;

                case sf::Keyboard::Up:
                    message.setCharacterSize( ++textSize );
                    break;

                case sf::Keyboard::Down:
                    message.setCharacterSize( --textSize );
                    break;
                }
            }
        }

        window.clear(sf::Color(50, 200, 50));
        window.draw(message);
        window.display();
    }

    return EXIT_SUCCESS;
}

```

### CODE REVIEW

File: src/SFML/Graphics/Font.cpp [hunk 1]
```diff
@@ -468,6 +468,26 @@ const Texture& Font::getTexture(unsigned int characterSize) const
     return m_pages[characterSize].texture;
 }
 
+////////////////////////////////////////////////////////////
+void Font::setSmooth( bool smooth )
```
Comment (eXpl0it3r) [inline src/SFML/Graphics/Font.cpp:472]: Remove spaces around argument

File: src/SFML/Graphics/Font.cpp [hunk 2]
```diff
@@ -468,6 +468,26 @@ const Texture& Font::getTexture(unsigned int characterSize) const
...
+{
+    if (smooth != m_isSmooth)
+    {
+        m_isSmooth = smooth;
+
+        for (int i = 0; i < m_pages.size(); ++i)
```
Comment (eXpl0it3r) [inline src/SFML/Graphics/Font.cpp:478]: Should use the key type `unsigned int` or probably even better, iterate over the container with iterators.

File: src/SFML/Graphics/Font.cpp [hunk 3]
```diff
@@ -468,6 +468,26 @@ const Texture& Font::getTexture(unsigned int characterSize) const
...
+        {
+            m_pages[i].texture.setSmooth( m_isSmooth );
```
Comment (eXpl0it3r) [inline src/SFML/Graphics/Font.cpp:480]: No spaces around the argument

File: include/SFML/Graphics/Font.hpp [hunk 4]
```diff
@@ -373,6 +399,7 @@ class SFML_GRAPHICS_API Font
     void*                      m_streamRec;   //!< Pointer to the stream rec instance (it is typeless to avoid exposing implementation details)
     void*                      m_stroker;     //!< Pointer to the stroker (it is typeless to avoid exposing implementation details)
     int*                       m_refCount;    //!< Reference counter used by implicit sharing
+    bool                       m_isSmooth;    //!< Status of the smooth filter
     Info                       m_info;        //!< Information about the font
```
Comment (eXpl0it3r) [inline include/SFML/Graphics/Font.hpp:403]: - Missing default initialization in constructor
- Missing copy in copy constructor
- Missing copy in assignment operator

**Review by eXpl0it3r — CHANGES_REQUESTED** [review]
@LaurentGomila mentioned on the linked [forum thread](https://en.sfml-dev.org/forums/index.php?topic=21058.msg150492#msg150492) that one can't just apply the smoothness like that. 
So the question is, if we simply change the texture property is this enough or do we need to re-render the already rendered glyphs?

**Comment by JonnyPtn** [comment]
Yeah he also clarified afterwards that he misunderstood and to ignore that comment, unless I misunderstand? the smoothing is not an issue when rendering glyphs, but when rendering texts

File: src/SFML/Graphics/Font.cpp (hunk 2 above)
Comment (JonnyPtn) [reply src/SFML/Graphics/Font.cpp]: Done, with iterators

File: src/SFML/Graphics/Font.cpp (hunk 3 above)
Comment (JonnyPtn) [reply src/SFML/Graphics/Font.cpp]: Done

File: include/SFML/Graphics/Font.hpp (hunk 4 above)
Comment (JonnyPtn) [reply include/SFML/Graphics/Font.hpp:403]: Done

File: src/SFML/Graphics/Font.cpp (hunk 1 above)
Comment (JonnyPtn) [reply src/SFML/Graphics/Font.cpp]: Done

**Comment by JonnyPtn** [comment]
Feedback addressed and rebased onto master

File: include/SFML/Graphics/Font.hpp [hunk 5]
```diff
@@ -275,6 +275,32 @@ class SFML_GRAPHICS_API Font
...
+    /// the font to look exactly the same as its source file,
+    /// you should disable it.
+    /// The smooth filter is enabled by default.
+    ///
+    /// \param smooth True to enable smoothing, false to disable it
+    ///
+    /// \see isSmooth
+    ///
+    ////////////////////////////////////////////////////////////
+    void setSmooth( bool smooth );
```
Comment (eXpl0it3r) [inline include/SFML/Graphics/Font.hpp:292]: Wrong spacing 😉

**Review by eXpl0it3r — APPROVED** [review]

File: include/SFML/Graphics/Font.hpp (hunk 5 above)
Comment (JonnyPtn) [reply include/SFML/Graphics/Font.hpp:292]: :( think It must be picking up a rogue clang-format file from somewhere as this is the style I use for other projects, will fix this though ( and out of curiosity would a clang-format PR be worthwhile? )

Comment (eXpl0it3r) [reply include/SFML/Graphics/Font.hpp:292]: I would very much appreciate such a PR. Not sure how easy it is to reproduce SFML's coding style, but I'd also be willing to change some of it in favor of an automated formatting.

Comment (JonnyPtn) [reply include/SFML/Graphics/Font.hpp]: Done

**Review by binary1248 — APPROVED** [review]

**Comment by binary1248** [comment]
SFMLBuildMaster: Build this please.

**Review by eXpl0it3r — APPROVED** [review]

**Pull request merged**